src/components/mongo.ts

```diff
@@ -0,0 +1,132 @@
+import * as pulumi from '@pulumi/pulumi';
+import * as aws from '@pulumi/aws';
+import * as awsx from '@pulumi/awsx';
+import { Ecs, RoleInlinePolicy } from './ecs-service';
+import { Size } from '../types/size';
+import { commonTags } from '../constants';
+
+export type MongoArgs = {
+  /**
+   * Exposed service port.
+   */
+  port: pulumi.Input<number>;
+  /**
+   * CPU and memory size used for running the container. Defaults to "small".
+   * Available predefined options are:
+   * - small (0.25 vCPU, 0.5 GB memory)
+   * - medium (0.5 vCPU, 1 GB memory)
+   * - large (1 vCPU memory, 2 GB memory)
+   * - xlarge (2 vCPU, 4 GB memory)
+   */
+  size?: pulumi.Input<Size>;
+  /**
+   * The aws.ecs.Cluster resource.
+   */
+  cluster: aws.ecs.Cluster;
+  /**
+   * The awsx.ec2.Vpc resource.
+   */
+  vpc: awsx.ec2.Vpc;
+  /**
+   * The environment variables to pass to a container. Don't use this field for
+   * sensitive information such as passwords, API keys, etc. For that purpose,
+   * please use the `secrets` property.
+   * Defaults to [].
+   */
+  environment?: aws.ecs.KeyValuePair[];
+  /**
+   * The secrets to pass to the container. Defaults to [].
+   */
+  secrets?: aws.ecs.Secret[];
+  taskExecutionRoleInlinePolicies?: pulumi.Input<
+    pulumi.Input<RoleInlinePolicy>[]
+  >;
+  taskRoleInlinePolicies?: pulumi.Input<pulumi.Input<RoleInlinePolicy>[]>;
+  /**
+   * A map of tags to assign to the resource.
+   */
+  tags?: pulumi.Input<{
+    [key: string]: pulumi.Input<string>;
+  }>;
+};
+
+export class Mongo extends pulumi.ComponentResource {
+  name: string;
+  service: Ecs;
+
+  constructor(
+    name: string,
+    args: MongoArgs,
+    opts: pulumi.ComponentResourceOptions = {},
+  ) {
+    super('studion:Mongo', name, args, opts);
+
+    const {
+      port,
+      size,
+      cluster,
+      vpc,
+      environment,
+      secrets,
+      taskExecutionRoleInlinePolicies,
+      taskRoleInlinePolicies,
+      tags,
+    } = args;
+
+    const securityGroup = new aws.ec2.SecurityGroup(
+      `${name}-service-security-group`,
+      {
+        vpcId: vpc.vpcId,
+        ingress: [
+          {
+            fromPort: port,
+            toPort: port,
+            protocol: 'tcp',
+            cidrBlocks: [vpc.vpc.cidrBlock],
+          },
+        ],
+        egress: [
+          {
+            fromPort: 0,
+            toPort: 0,
+            protocol: '-1',
+            cidrBlocks: ['0.0.0.0/0'],
+          },
+        ],
+        tags: commonTags,
+      },
+      { parent: this },
+    );
+
+    this.name = name;
+    this.service = new Ecs(
+      name,
+      {
+        image:
+          'mongo@sha256:238b1636bdd7820c752b91bec8a669f92568eb313ad89a1fc4a92903c1b40489',
+        port,
+        cluster,
+        desiredCount: 1,
+        minCount: 1,
+        maxCount: 1,
+        ...(size && { size }),
```

Comment:
```suggestion
       size,
```
This doesn't work?

Reply:
No, it does not.
If I provide `size`, and user does not enter any size the value will be `undefined`.
`argsWithDefaults` would then set the size arg to `undefined` and that value would result in error.

<img width="386" alt="image" src="https://github.com/ExtensionEngine/infra-code-blocks/assets/139864784/d30d4e36-cdd9-4afb-9c37-4e57394f21c0">

Reply:
You can just spread args instead of destructing each property from args and then manually applying it to EcsService  args:
```ts
  this.service = new Ecs(
      name,
      {
        ...args,
        image:
          'mongo@sha256:238b1636bdd7820c752b91bec8a669f92568eb313ad89a1fc4a92903c1b40489',
        desiredCount: 1,
        minCount: 1,
        maxCount: 1,
        enableServiceAutoDiscovery: true,
        persistentStorageVolumePath: '/data/db',
        dockerCommand: ['mongod', '--port', port.toString()],
        assignPublicIp: false,
        securityGroup,
      },
      { ...opts, parent: this },
    );
```
By doing so, you are only passing arguments that are specified by the consumer and in that case `size` won't be passed to ecs service if it's not set by consumer